src/CommonLib/LDAPUtils.cs

```diff
@@ -1415,163 +1438,534 @@ private static bool RequestNETBIOSNameFromComputer(string server, string domain,
+                else
+                {
+                    CachedDomainInfo.TryGetValue(domain, out info);
+                }
+                return new LdapConnectionWrapper
+                {
+                    Connection = connection,
+                    DomainInfo = info
+                };
+            }
+
+            CheckAndThrowException(connectionResult.Exception);
+
+            //If we're not allowing fallbacks to LDAP from LDAPS, just return here
+            if (_ldapConfig.ForceSSL)
             {
-                var cred = new NetworkCredential(_ldapConfig.Username, _ldapConfig.Password);
-                connection.Credential = cred;
+                return null;
             }
+            //If we get to this point, it means we have an unsuccessful connection, but our error code doesn't indicate an outright failure
+            //Try a new connection without SSL
+            connection = CreateConnectionHelper(target, false, authType, globalCatalog);
 
-            //These options are important!
-            connection.SessionOptions.ProtocolVersion = 3;
-            connection.SessionOptions.ReferralChasing = ReferralChasingOptions.None;
+            connectionResult = TestConnection(connection);
+                
+            if (connectionResult.Success)
+            {
+                var domain = connectionResult.DomainInfo.DomainFQDN;
+                if (!CachedDomainInfo.ContainsKey(domain.ToUpper()))
+                {
+                    var baseDomainInfo = connectionResult.DomainInfo;
+                    baseDomainInfo.DomainSID =  GetDomainSid(connection, baseDomainInfo);
+                    baseDomainInfo.DomainNetbiosName = GetDomainNetbiosName(connection, baseDomainInfo);
+                    CachedDomainInfo.TryAdd(baseDomainInfo.DomainFQDN, baseDomainInfo);
+                    CachedDomainInfo.TryAdd(baseDomainInfo.DomainNetbiosName, baseDomainInfo);
+                    CachedDomainInfo.TryAdd(baseDomainInfo.DomainSID, baseDomainInfo);
+                    
+                    if (!string.IsNullOrEmpty(baseDomainInfo.DomainSID))
+                    {
+                        Cache.AddDomainSidMapping(baseDomainInfo.DomainFQDN, baseDomainInfo.DomainSID);
+                    }
 
-            if (_ldapConfig.DisableSigning)
+                    if (!string.IsNullOrEmpty(baseDomainInfo.DomainNetbiosName))
+                    {
+                        Cache.AddDomainSidMapping(baseDomainInfo.DomainNetbiosName, baseDomainInfo.DomainSID);
+                    }
+
+                    info = baseDomainInfo;
+                }else
+                {
+                    CachedDomainInfo.TryGetValue(domain, out info);
+                }
+                return new LdapConnectionWrapper
+                {
+                    Connection = connection,
+                    DomainInfo = info
+                };
+            }
+            
+            CheckAndThrowException(connectionResult.Exception);
```

Comment:
Should this be moved up higher in the function? There's a lot of logic between acquiring `connectionResult` and checking the `Exception` field.

Reply:
The in between logic will only apply if we have a successful connection, and if we do we return before hitting this logic. It doesn't make any functional difference to the logic of this function, since success conditions never generate exceptions